### ISSUE TITLE
Inconsistent doc_count in composite aggregations when applied on multiple multi-valued fields

### ISSUE THREAD

**Issue by Luegg** [elastic/elasticsearch#88598]
### Elasticsearch Version

8.4.0-SNAPSHOT

### Java Version

_bundled_

### OS Version

OS X

### Problem Description

Given an index with one doc:
```
{"a": [2,1], "b": [1,2], "c": [1,1], "d": [1,1]}
```

I’m using the following query to perform composite aggs on two fields (and sum a third one):

```
{
  "size": 0,
  "aggs": {
      "ac":{
        "composite": {
            "sources": [
                {"a": {"terms": {"field": "a"}}},
                {"c": {"terms": {"field": "c"}}}
            ]
        },
        "aggs": {
            "s": {"sum": {"field": "b"}}
        }
      },
      "cd":{
        "composite": {
            "sources": [
                {"c": {"terms": {"field": "c"}}},
                {"d": {"terms": {"field": "d"}}}
            ]
        },
        "aggs": {
            "s": {"sum": {"field": "b"}}
        }
      }
  }
}
```

Now the returned result is:

```
{
  ...
  "aggregations": {
    "cd": {
      ...
      "buckets": [
        {
          "key": {"c": 1, "d": 1},
          "doc_count": 2, // <== should probably be 4
          "s": {"value": 12.0}
        }
      ]
    },
    "ac": {
      ...
      "buckets": [
        {
          "key": {"a": 1, "c": 1},
          "doc_count": 2,
          "s": {"value": 6.0}
        },
        {
          "key": {"a": 2, "c": 1},
          "doc_count": 2,
          "s": {"value": 6.0}
        }
      ]
    }
  }
}
```

This comes at a surprise to me because I’d have expected the total doc_counts to be equal for both aggregations (4) but it’s only 2 for cd. Note also, that the total sums add up to the same value.

As an additional observation, this behavior only appears if `missing_bucket` is missing or `false` for the value sources. Hence, one can work around the issue by using

```
      "cd":{
        "composite": {
            "sources": [
                {"c": {"terms": {"field": "c","missing_bucket": true}}},
                {"d": {"terms": {"field": "d","missing_bucket": true}}}
            ]
        },
        "aggs": {
            "s": {"sum": {"field": "b"}}
        }
      }
```

### Steps to Reproduce

see above

**Comment by ywelsch** [comment]
Another example showing the discrepancies between terms aggregation and composite terms aggregation when it comes to handling multi-value fields, with a more minimal example:


```
DELETE test

PUT test/_doc/id1?refresh=true
{
  "foo" : [1, 1]
}

GET test/_search
{
  "docvalue_fields": ["foo"],
  "aggs": {
    "terms_foo": {
      "terms": {
        "field": "foo"
      },
      "aggs": {
        "stats": {
          "sum": {
            "field": "foo"
          }
        }
      }
    },
    "comp_foo": {
      "composite": {
        "sources": [
          {
            "foo": {
              "terms": {
                "field": "foo"
              }
            }
          }
        ]
      },
      "aggs": {
        "stats": {
          "sum": {
            "field": "foo"
          }
        }
      }
    }
  }
}
```

which yields

```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test",
        "_id": "id1",
        "_score": 1,
        "_source": {
          "foo": [
            1,
            1
          ]
        },
        "fields": {
          "foo": [
            1,
            1
          ]
        }
      }
    ]
  },
  "aggregations": {
    "comp_foo": {
      "after_key": {
        "foo": 1
      },
      "buckets": [
        {
          "key": {
            "foo": 1
          },
          "doc_count": 1,
          "stats": {
            "value": 4
          }
        }
      ]
    },
    "terms_foo": {
      "doc_count_error_upper_bound": 0,
      "sum_other_doc_count": 0,
      "buckets": [
        {
          "key": 1,
          "doc_count": 1,
          "stats": {
            "value": 2
          }
        }
      ]
    }
  }
}
```

**Comment by not-napoleon** [comment]
@ywelsch can we close this now that #88638 has merged?

**Comment by ywelsch** [comment]
The example I gave has been fixed by #88638.

For the example that @Luegg provided, the output is now independent of missing_bucket and yields:

```
{
  "took": 9,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "cd": {
      "after_key": {
        "c": 1,
        "d": 1
      },
      "buckets": [
        {
          "key": {
            "c": 1,
            "d": 1
          },
          "doc_count": 1,
          "s": {
            "value": 3
          }
        }
      ]
    },
    "ac": {
      "after_key": {
        "a": 2,
        "c": 1
      },
      "buckets": [
        {
          "key": {
            "a": 1,
            "c": 1
          },
          "doc_count": 1,
          "s": {
            "value": 3
          }
        },
        {
          "key": {
            "a": 2,
            "c": 1
          },
          "doc_count": 1,
          "s": {
            "value": 3
          }
        }
      ]
    }
  }
}
```

This looks ok to me now given the "set" interpretation for multi-valued field on grouping.

If @Luegg is happy, we can close the issue.

**Comment by Luegg** [comment]
Thanks a lot @ywelsch! I'm even more happy about the clarification of the intended behavior :)
